diff --git a/nols_website/wmi/static/wmi/js/dataentry.js b/nols_website/wmi/static/wmi/js/dataentry.js
index b82308af7..802c9418f 100644
--- a/nols_website/wmi/static/wmi/js/dataentry.js
+++ b/nols_website/wmi/static/wmi/js/dataentry.js
@@ -1,64 +1,52 @@
 $('.modal-form').each(function(){
-  $(this).submit(function(e) {
+  var form = $(this)
+  form.submit(function(e) {
       e.preventDefault();
-      url = $(this).attr('data-form-url');
-      return validateCoreForm($(this), url);
-      // {% ifequal form_id 'phone_form' %}
-      // var inputs = $(this).find('[name*="phoneform-"]')
-      // for(var i=0; i<inputs.length; i++) {
-      //     var new_name = $(inputs[i]).attr('name').replace(/phoneform-/, '')
-      //     $(inputs[i]).attr('name', new_name)
-      // }
-      // {% endifequal %}
+      var url = form.attr('data-form-url');
+      $.ajax({
+          type: "POST",
+          url: url,
+          data: form.serialize().replace(/phoneform-/g, '')
+      }).done(function(response) {
+        if (response.is_valid){
+          form.unbind('submit').submit();
+        } else {
+          showErrors(form, response);
+        }
+      });
   });
 });
-function validateCoreForm(core_form, url) {
-    var valid = false;
-
-    var PREFIX = 'phoneform'
-    var prefix = '';
-    if(core_form.parent().attr('id') == 'phone_form') {
-        prefix = PREFIX + '-';
-    }
 
-    $.ajax({
-        type: "POST",
-        url: url,
-        data: core_form.serialize().replace(/phoneform-/g, '')
-    }).done( function( data ) {
-        if(data.is_valid) {
-            valid = true;
-        } else {
-            core_form.find('.form-group').removeClass('error');
-            core_form.find('.errorlist').parent('div').remove();
-            core_form.find('.group-errors').remove();
-            if(typeof data == 'string') {
-                // display the authlockout page
-                $('html').replaceWith(data);
-            } else {
-                for(var field in data) {
-                    if(field != '__all__') {
-                        var id = '#id_' + prefix + field;
-                        var formGroup = $(id).parents('.form-group');
-                        formGroup.addClass('error');
-                        formGroup.prepend('<div class="row"><ul class="errorlist column medium-6 column medium-offset-3 fade"><li>'
-                            + data[field] + '</li></ul></div>')
-                    } else {
-                        core_form.find('.form-instructions').after(
-                            '<div class="form-group error group-errors"><ul class="errorlist fade"><li>'
-                                + data[field] + '</li></ul></div> '
-                        );
-                    }
-                }
-            }
-            valid = false;
-        }
-    });
-    window.setTimeout(function(){
-        $('.errorlist.fade').addClass('in');
-    }, 100);
-    if(valid){
-        core_form.find('input[type="submit"]').attr('disabled', 'disabled');
-    }
-    return valid;
+function showErrors(form, response){
+  var PREFIX = 'phoneform'
+  var prefix = '';
+  //------------------------------------
+  //TODO need to tweak this if necessary
+  //------------------------------------
+  if(form.attr('id') == 'phone_form') {
+      prefix = PREFIX + '-';
+  }
+  form.find('.form-group').removeClass('error');
+  form.find('.errorlist').parent('div').remove();
+  form.find('.group-errors').remove();
+  if(typeof response == 'string') {
+      // display the authlockout page
+      $('html').replaceWith(response);
+  } else {
+      for(var field in response) {
+          if(field != '__all__') {
+              var id = '#id_' + prefix + field;
+              var formGroup = $(id).parents('.form-group');
+              formGroup.addClass('error');
+              formGroup.prepend('<div class="row"><ul class="errorlist column medium-6 column medium-offset-3 fade"><li>'
+                  + response[field] + '</li></ul></div>')
+          } else {
+              form.find('.form-instructions').after(
+                  '<div class="form-group error group-errors"><ul class="errorlist fade"><li>'
+                      + response[field] + '</li></ul></div> '
+              );
+          }
+      }
+  }
+  $('.errorlist.fade').addClass('in');
 }
