diff --git a/nols_website/core/contextproc/dataentry.py b/nols_website/core/contextproc/dataentry.py
index c0840a3be..51fda5705 100644
--- a/nols_website/core/contextproc/dataentry.py
+++ b/nols_website/core/contextproc/dataentry.py
@@ -1,9 +1,8 @@
-# from django.contrib.auth.forms import AuthenticationForm
+from django.contrib.auth.forms import AuthenticationForm
 
 
 def auth_form(request):
+    if 'rollcall' in request.path and not request.user.is_authenticated():
+        request.session.set_test_cookie()
+        return {'auth_form': AuthenticationForm(request=request)}
     return {}
-    # if 'rollcall' in request.path and not request.user.is_authenticated():
-    #     request.session.set_test_cookie()
-    #     return {'auth_form': AuthenticationForm(request=request)}
-    # return {}
